pkg/cluster/clustermsi.go

```diff
@@ -0,0 +1,145 @@
+package cluster
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore"
+	"github.com/Azure/azure-sdk-for-go/sdk/azcore/arm"
+	"github.com/Azure/msi-dataplane/pkg/dataplane"
+	"github.com/Azure/msi-dataplane/pkg/store"
+
+	"github.com/Azure/ARO-RP/pkg/env"
+	"github.com/Azure/ARO-RP/pkg/util/azureclient/azuresdk/armmsi"
+)
+
+const (
+	mockMsiCertValidityDays = 90
+)
+
+// ensureClusterMsiCertificate leverages the MSI dataplane module to fetch the MSI's
+// backing certificate (if needed) and store the certificate in the cluster MSI key
+// vault. It does not concern itself with whether an existing certificate is valid
+// or not; that can be left to the certificate refresher component.
+func (m *manager) ensureClusterMsiCertificate(ctx context.Context) error {
+	secretName, err := m.clusterMsiSecretName()
+	if err != nil {
+		return err
+	}
+
+	_, err = m.clusterMsiKeyVaultStore.GetCredentialsObject(ctx, secretName)
+	if err == nil {
+		return nil
+	} else if azcoreErr, ok := err.(*azcore.ResponseError); !ok || azcoreErr.StatusCode != http.StatusNotFound {
+		return err
+	}
+
+	clusterMsiResourceId, err := m.doc.OpenShiftCluster.ClusterMsiResourceId()
+	if err != nil {
+		return err
+	}
+
+	uaMsiRequest := dataplane.UserAssignedMSIRequest{
+		IdentityURL: m.doc.OpenShiftCluster.Identity.IdentityURL,
+		ResourceIDs: []string{clusterMsiResourceId.String()},
+		TenantID:    m.doc.OpenShiftCluster.Identity.TenantID,
+	}
+
+	msiCredObj, err := m.msiDataplane.GetUserAssignedIdentities(ctx, uaMsiRequest)
+	if err != nil {
+		return err
+	}
+
+	now := time.Now()
+
+	var expirationDate time.Time
+	if m.env.FeatureIsSet(env.FeatureUseMockMsiRp) {
+		expirationDate = now.AddDate(0, 0, mockMsiCertValidityDays)
+	} else {
+		if msiCredObj.CredentialsObject.ExplicitIdentities == nil || len(msiCredObj.CredentialsObject.ExplicitIdentities) == 0 || msiCredObj.CredentialsObject.ExplicitIdentities[0] == nil || msiCredObj.CredentialsObject.ExplicitIdentities[0].NotAfter == nil {
```

Comment:
Similarly to @jaitaiwan 's comment, this is pretty long and could be a separate function, but it's just a matter of taste I think.

Reply:
I think I'm gonna skip making a function for this one if that's cool, since I don't foresee us needing to check a `CredentialsObject`'s `NotAfter` outside of this function.